File: check-proxy.rb
#!/opt/homebrew/opt/ruby@3.3/bin/ruby
# frozen_string_literal: true

#
# Check proxy connectivity
# Tests HTTP connection via specified proxy server.
# Uses standard library only (open3, optparse).
#
# Usage:
#   check-proxy.rb [options]
#
# Options:
#   -d, --debug           Show debug info
#   -p, --proxy URL       Set proxy server URL
#   -n, --name NAME       Proxy server name from built-in list
#   -l, --list            Show known proxies
#   -i, --isp             Check which ISP the connection uses
#   -h, --help            Show this help
#

require 'open3'
require 'optparse'

PROXY_SERVERS = [
  { name: 'none',   url: 'none' },                          # no proxy
  { name: 'socks',  url: 'socks://nuc.local:3228' },        # nuc.local socks
  { name: 'squid',  url: 'http://nuc.local:3128' },         # nuc.local squid
  { name: 'local',  url: 'http://localhost:3128' },         # localhost squid
  { name: 'office', url: 'http://proxy.iiji.jp:8080' },     # IIJ Office proxy
].freeze

TEST_TARGET   = 'https://www.google.com'
DETAIL_TARGET = 'https://env.b4iine.net'
TEMP_BODY     = '/tmp/curl.out'

def parse_isp_from_html(path)
  html = File.read(path, encoding: 'utf-8')
  # Locate first env_1box block, then extract txt content (strip inner tags)
  start = html.index('<div class="env_1box">')
  raise "env_1box not found in #{path}" unless start

  txt_match = html[start..].match(/<div[^>]+class="txt"[^>]*>(.*?)<\/div>/m)
  raise "txt not found in #{path}" unless txt_match

  txt_match[1].gsub(/<[^>]+>/, '').strip
rescue StandardError => e
  raise "error on reading from #{path}: #{e}"
end

def main
  options = { debug: false, proxy: nil, name: nil, list: false, isp: false }

  OptionParser.new do |opts|
    opts.banner = 'Usage: check-proxy.rb [options]'
    opts.separator ''
    opts.separator 'Options:'

    opts.on('-d', '--debug',        'Show debug info')            { options[:debug] = true }
    opts.on('-p', '--proxy URL',    'Set proxy server URL')       { |v| options[:proxy] = v }
    opts.on('-n', '--name NAME',    'Proxy server name from list',
            PROXY_SERVERS.map { |s| s[:name] })                   { |v| options[:name] = v }
    opts.on('-l', '--list',         'Show known proxies')         { options[:list] = true }
    opts.on('-i', '--isp',          'Check which ISP is in use')  { options[:isp] = true }
    opts.on_tail('-h', '--help',    'Show this help')             { puts opts; exit 0 }
  end.parse!

  if options[:list]
    PROXY_SERVERS.each { |s| puts "%-8s %s" % [s[:name], s[:url]] }
    return
  end

  proxy_url = if options[:name]
    entry = PROXY_SERVERS.find { |s| s[:name] == options[:name] }
    $stderr.puts "Unknown proxy name: #{options[:name]}" unless entry
    entry&.fetch(:url)
  elsif options[:proxy]
    options[:proxy]
  else
    PROXY_SERVERS.first[:url]
  end

  debug_source = options[:name] ? 'name option' : (options[:proxy] ? 'proxy option' : 'default')
  puts "Set proxy to #{proxy_url} by #{debug_source}" if options[:debug]

  target_url = options[:isp] ? DETAIL_TARGET : TEST_TARGET
  body_out   = options[:isp] ? TEMP_BODY     : '/dev/null'

  puts "Trying #{target_url} via proxy #{proxy_url} ... " if options[:debug]

  proxy_args = proxy_url == 'none' ? [] : ['--proxy', proxy_url]
  cmd = ['/usr/bin/curl', '--connect-timeout', '5', '-s',
         *proxy_args, '-o', body_out, '-w', '%{http_code}\n', target_url]

  output, status = Open3.capture2e(*cmd)
  code = status.success? ? output.lines.first&.strip.to_i : 0

  if code == 200
    if options[:isp]
      begin
        isp = parse_isp_from_html(TEMP_BODY)
        puts "OK/#{isp}"
      rescue StandardError => e
        puts e.message
      end
    else
      puts 'OK'
    end
  else
    puts "NG (#{code})"
  end
end

main if $PROGRAM_NAME == __FILE__

File: check-proxy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import subprocess
import argparse
from bs4 import BeautifulSoup
from pprint import pprint


proxy_servers = [
    {'name':'none',   'url':"none"},    # no proxy
    {'name':'socks',  'url':"socks://nuc.local:3228"},    # nuc.local socks
    {'name':'squid',  'url':"http://nuc.local:3128"},     # nuc.local squid
    {'name':'local',  'url':"http://localhost:3128"},     # localhost squid
    {'name':'office', 'url':"http://proxy.iiji.jp:8080"}, # IIJ Office proxy
]

proxy_servers_names = [item["name"] for item in proxy_servers]

test_target = "https://www.google.com"
detail_target = "https://env.b4iine.net"
temp_body = "/tmp/curl.out"

def main():
    global test_target

    parser = argparse.ArgumentParser(description="Check socks status")
    parser.add_argument("-d", "--debug", action="store_true", help="use debug")
    parser.add_argument("-p", "--proxy", default=None, help="set proxy server")
    parser.add_argument("-n", "--name", choices=proxy_servers_names, help="proxy server in lists")
    parser.add_argument("-l", "--list", action="store_true", help="show known proxies")
    parser.add_argument("-i", "--isp", action="store_true", help="check connection which isp")
#    parser.add_argument("-t", "--target", default=test_target, help="test target url")
    args = parser.parse_args()

    if args.list :
        pprint(proxy_servers)
        return 0

    if args.name : 
        result = [item for item in proxy_servers if item["name"] == args.name]
        if len(result) > 0 :
            proxy_server = result[0]['url']
            if args.debug :
                print(f"Set proxy to {proxy_server} by name option")
    else :
        if args.proxy is not None:
            proxy_server = args.proxy
            if args.debug :
                print(f"Set proxy to {proxy_server} by proxy option")
        else :
            proxy_server = proxy_servers[0]['url']
            if args.debug :
                print(f"Set proxy to {proxy_server} by default")

    code = 0

    if args.isp : 
        target_url = detail_target
        body_out = temp_body
    else :
        target_url = test_target
        body_out = "/dev/null"

    if args.debug :
        print(f"Trying {target_url} via proxy {proxy_server} ... ", end="")

    if proxy_server == "none" :
        proxy_str = ""
    else : 
        proxy_str = f"--proxy {proxy_server}"

    check_cmd = f"/usr/bin/curl --connect-timeout 5 -s {proxy_str} -o {body_out} -w '%{{http_code}}\n' {target_url}"

    try:
        output = subprocess.check_output(check_cmd, shell=True, stderr=subprocess.STDOUT)
        code = int(output.decode().strip().split("\n").pop(0))

    except subprocess.CalledProcessError as e:
        print(e)
    finally:
        if code == 200:
            if args.isp : 
                try:
                    with open(temp_body, 'r', encoding='utf-8') as f:
                        soup = BeautifulSoup(f.read(), "html.parser")
                        env_1box = soup.find('div', class_='env_1box')
                        isp = env_1box.find('div', class_='txt').text.strip()
                        print(f"OK/{isp}")
                except Exception as e:
                    print(f"error on reading fou {temp_body}: {e}")
            else :
                print("OK")
        else:
            print(f"NG ({code})")

    return 0

if __name__ == "__main__":
    main()
